### PR TITLE
[ldapjs-client] Basic Typings

### PR DESCRIPTION
**Pull request by colthreepv**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

# ldapjs-client
Those typings are not the best they can get, as they are inferred by my use of the library, and my ability to read the original source code.

All of the `Promise<any>` that you see are quite correct, the response from LDAP servers is implementation dependent.

I have met a weird tslint error, it kept telling me the following line had unnecessary generics
```typescript
search<T extends any>(base: string, options: PartialSearchOptions): Promise<T[]>;
```

But without (an optional) generic the user cannot specify what kind of response it's going to get. I find that preferable instead of using a type cast after having received the data.

Also I had to disable "npm-naming" in tslint because it kept arguing the original library has no "default" in it, but it just uses `module.exports`. I can't use `module.exports` in Typescript, of course.

Any Insight would be appreciated, that's my first added typing

### CODE REVIEW

File: types/ldapjs-client/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+     */
+    modifyDN(dn: string, newName: string): Promise<any>;
+
+    /**
+     * Performs an LDAP search against the server.
+     *
+     * Note that the defaults for options are a 'base' search.
+     */
+    // tslint:disable-next-line: no-unnecessary-generics
+    search<T extends any>(base: string, options: PartialSearchOptions): Promise<T[]>;
```
Comment (DanielRosenwasser) [inline types/ldapjs-client/index.d.ts]: I think this is legitimate - the generic is really acting as a type assertion. You should instead pick `unknown[]` and let users cast.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

File: types/ldapjs-client/tslint.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+    "extends": "dtslint/dt.json",
+    "rules": {
+        "npm-naming": false
```
Comment (DanielRosenwasser) [inline types/ldapjs-client/tslint.json]: Why is this needed?

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

File: types/ldapjs-client/tslint.json (hunk 2 above)
Comment (colthreepv) [reply types/ldapjs-client/tslint.json]: I wrote it in the PR description, to be more verbose, it gives the following error:
```
npm-naming  The types for ldapjs-client specify 'export default' but the source does not mention 'default' anywhere.
```

The sources of ldapjs-client have `module.exports = Client;`, but I have to use `export default LdapClient` on the typings

File: types/ldapjs-client/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    search(base: string, options: PartialSearchOptions): Promise<unknown[]>;
+
+    /**
+     * Unbinds this client from the LDAP server.
+     *
+     */
+    unbind(): Promise<any>;
+}
+
+export default LdapClient;
```
Comment (DanielRosenwasser) [inline types/ldapjs-client/index.d.ts]: This package was declared using `module.exports = ...` or some variant. It should be declared with an `export = ...` like one of the following:

### Variables

```ts
declare var xyz: SomeType;

export = xyz;
```

### Classes

```ts
class Foo {
  // ...
}

export = Foo
```

### Functions

```ts
function foo(): any;

export = foo;
```

To import the entity using a default import, you will need to enable `esModuleInterop`. Alternatively, you can just import using a more appropriate syntax like

```ts
import foo = require("foo");
```

[See more here in the FAQ](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/9f9f6430117d025aef62ed56128deb65508edc86/README.md#a-package-uses-export--but-i-prefer-to-use-default-imports-can-i-change-export--to-export-default).

Comment (colthreepv) [reply types/ldapjs-client/index.d.ts]: Ok perfect. Sorry for have missed that part.

There is a problem though, I don't seem to be able to export interfaces and types alongside the default export being LdapClient

Is there a way in Typescript to do this? I have googled but not many people use commonjs exports  in the wild (outside definitelytyped)

**Comment by colthreepv** [comment]
Ok, I have re-enabled the `npm-naming` rule, still my question stands, if anyone more expert than me on TS could answer

> There is a problem though, I don't seem to be able to export interfaces and types alongside the default export being LdapClient
> 
> Is there a way in Typescript to do this? I have googled but not many people use commonjs exports in the wild (outside definitelytyped)

File: types/ldapjs-client/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
+// Type definitions for ldapjs-client 0.1
+// Project: https://github.com/zont/ldapjs-client#readme
+// Definitions by: Valerio Coltrè <https://github.com/colthreepv>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 3.0
+
+declare class Filter {
```
Comment (DanielRosenwasser) [inline types/ldapjs-client/index.d.ts]: Is `Filter` actually a class, or is it an interface?

File: types/ldapjs-client/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+interface Change {
+    operation: string;
+    modification: {
+        [key: string]: any;
+    };
+}
+
+interface SearchOptions {
```
Comment (DanielRosenwasser) [inline types/ldapjs-client/index.d.ts]: If this is only used in one place, I'd really just avoid writing two types.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]
If you really want to be able to reference the types too, you'll need to do something called namespace merging in which you write something like

```
class LdapClient {
    // ...
}

namespace LdapClient {
    export interface Change {
      // ...
    }
}

File: types/ldapjs-client/index.d.ts (hunk 4 above)
Comment (colthreepv) [reply types/ldapjs-client/index.d.ts]: I have never used that.
Blatantly copied from [ldapjs](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/ldapjs/index.d.ts#L354)

Honestly [reading carefully the source code](https://github.com/zont/ldapjs-client/blob/master/src/index.js#L96), I'm going to remove this altogether, and just use `string`

Even if the ldapjs-client library has an [entire directory devoted to support Class-Filters](https://github.com/zont/ldapjs-client/tree/master/src/filters)

File: types/ldapjs-client/index.d.ts (hunk 5 above)
Comment (colthreepv) [reply types/ldapjs-client/index.d.ts]: sure

**Comment by colthreepv** [comment]
Thanks a lot for this insight, clever use of merging indeed!
I have refined the definitions again

**Review by colthreepv — COMMENTED** [review]
Thanks a lot for this insight, clever use of merging indeed!
I have refined the definitions again

**Comment by colthreepv** [comment]
@DanielRosenwasser sorry I think my forced-update broke the PR change request flow
can you review again?

**Comment by DanielRosenwasser** [comment]
Thanks!
